src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+        }
+
+        /// <summary>
+        /// Fetches and processes the current document model.
+        /// </summary>
+        private async ValueTask GetModelAsync(CancellationToken cancellationToken)
+        {
+            await TaskScheduler.Default;
+            CodeWindow.GetLastActiveView(out var textView);
+            var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+            if (activeTextView is not null)
+            {
+                // Need to be on UI thread to get the file path.
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var filePath = GetFilePath(activeTextView);
+
+                // Ensure we switch to the threadpool before calling DocumentSymbolsRequestAsync.  It ensures
+                // that fetching and processing the document model is not done on the UI thread.
+                await TaskScheduler.Default;
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    activeTextView.TextBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is not null)
+                {
+                    var responseBody = response.ToObject<DocumentSymbol[]>();
+                    var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                    LatestFetchedDocumentModel = DocumentOutlineHelper.GetDocumentSymbolModels(documentSymbols);
+
+                    LspSnapshot = activeTextView.TextSnapshot;
+                    LatestFetchedDocumentModelIsNotEmpty = LatestFetchedDocumentModel.Length > 0;
+
+                    StartModelUpdateTask();
+                }
+                else
+                {
+                    LatestFetchedDocumentModelIsNotEmpty = false;
+                    LatestFetchedDocumentModel = ImmutableArray<DocumentSymbolViewModel>.Empty;
+                }
+            }
+            else
+            {
+                LatestFetchedDocumentModelIsNotEmpty = false;
+                LatestFetchedDocumentModel = ImmutableArray<DocumentSymbolViewModel>.Empty;
+            }
+
+            string? GetFilePath(IWpfTextView textView)
+            {
+                ThreadingContext.ThrowIfNotOnUIThread();
+                if (textView.TextBuffer.Properties.TryGetProperty(typeof(IVsTextBuffer), out IVsTextBuffer bufferAdapter) &&
+                    bufferAdapter is IPersistFileFormat persistFileFormat &&
+                    ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+                {
+                    return filePath;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Processes the fetched document model and updates the UI.
+        /// </summary>
+        private async ValueTask UpdateModelAsync(CancellationToken cancellationToken)
```

Comment:
i think you may be using 'model' to refer to two separate things (please correct me if i'm wrong).  but it soundsl ike you're using it to reference the data that LSP returns *And* to reference the wpf UI you're building.  THat's a bit confusing reading these methods as it's unclear which you're referring to.

Reply:
I've updated the wording to make it clearer that model refers to the data that LSP returns.

Reply:
I have renamed some methods, please let me know if that's more clear!